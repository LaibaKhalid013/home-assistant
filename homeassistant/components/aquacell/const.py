"""Constants for the Aquacell integration."""

DOMAIN = "aquacell"
DATA_AQUACELL = "DATA_AQUACELL"
