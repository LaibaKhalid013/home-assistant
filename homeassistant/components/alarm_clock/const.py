"""Constants for the Alarm Clock integration."""

import voluptuous as vol

from homeassistant.const import CONF_ICON, CONF_NAME
import homeassistant.helpers.config_validation as cv

DOMAIN = "alarm_clock"
ENTITY_ID_FORMAT = DOMAIN + ".{}"

ATTR_ALARM_TIME = "alarm_time"
ATTR_REPEAT_DAYS = "repeat_days"
ATTR_NEXT_ALARM = "next_alarm"

CONF_ALARM_TIME = "alarm_time"
CONF_REPEAT_DAYS = "repeat_days"

EVENT_ALARM_CLOCK_STARTED = "alarm_clock.started"
EVENT_ALARM_CLOCK_FINISHED = "alarm_clock.finished"
EVENT_ALARM_CLOCK_CANCELLED = "alarm_clock.cancelled"
EVENT_ALARM_CLOCK_CHANGED = "alarm_clock.changed"

STORAGE_KEY = DOMAIN
STORAGE_VERSION = 1

STORAGE_FIELDS = {
    vol.Required(CONF_NAME): cv.string,
    vol.Optional(CONF_ICON): cv.icon,
    vol.Required(CONF_ALARM_TIME): cv.time,
    vol.Required(CONF_REPEAT_DAYS): cv.weekdays,
}
