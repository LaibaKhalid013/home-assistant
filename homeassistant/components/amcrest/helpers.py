"""Helpers for amcrest component."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.helpers.typing import UndefinedType


def service_signal(service: str, *args: str) -> str:
    """Encode signal."""
    return "_".join([DOMAIN, service, *args])


def log_update_error(
    logger: logging.Logger,
    action: str,
    name: str | UndefinedType | None,
    entity_type: str,
    error: Exception,
    level: int = logging.ERROR,
) -> None:
    """Log an update error."""
    logger.log(
        level,
        "Could not %s %s %s due to error: %s",
        action,
        name,
        entity_type,
        error.__class__.__name__,
    )
