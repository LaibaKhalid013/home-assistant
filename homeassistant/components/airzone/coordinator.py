"""The Airzone integration."""

from __future__ import annotations

from asyncio import timeout
from datetime import timedelta
import logging
from typing import Any

from aioairzone.const import AZD_NEW_SYSTEMS, AZD_NEW_ZONES
from aioairzone.exceptions import AirzoneError
from aioairzone.localapi import AirzoneLocalApi

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import AIOAIRZONE_DEVICE_TIMEOUT_SEC, DOMAIN

SCAN_INTERVAL = timedelta(seconds=60)

_LOGGER = logging.getLogger(__name__)


class AirzoneUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
    """Class to manage fetching data from the Airzone device."""

    def __init__(self, hass: HomeAssistant, airzone: AirzoneLocalApi) -> None:
        """Initialize."""
        self.airzone = airzone

        super().__init__(
            hass,
            _LOGGER,
            name=DOMAIN,
            update_interval=SCAN_INTERVAL,
        )

    async def _async_check_new_devices(self, data: dict[str, Any]) -> None:
        """Reload entry if there are new devices."""
        if (
            len(data.get(AZD_NEW_SYSTEMS, [])) > 0
            or len(data.get(AZD_NEW_ZONES, [])) > 0
        ):
            assert self.config_entry
            self.hass.config_entries.async_schedule_reload(self.config_entry.entry_id)

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        async with timeout(AIOAIRZONE_DEVICE_TIMEOUT_SEC):
            try:
                await self.airzone.update()
            except AirzoneError as error:
                raise UpdateFailed(error) from error
            data = self.airzone.data()
        await self._async_check_new_devices(data)
        return data
