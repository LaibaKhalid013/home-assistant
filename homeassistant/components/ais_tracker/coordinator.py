"""AIS tracker coordinator."""

from __future__ import annotations

import asyncio
import logging
from socket import AF_INET, SO_REUSEPORT, SOCK_DGRAM, SOL_SOCKET, socket

from pyais.stream import SocketStream

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PORT
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
from homeassistant.util.json import json_loads_object

from .const import CONF_MMSIS, DOMAIN

LOGGER = logging.getLogger(__name__)

type AisTrackerConfigEntry = ConfigEntry[AisTrackerCoordinator]


class UDPReceiver(SocketStream):
    """Re-implementation of pyais.stream.UDPReceiver.

    can be removed, as soon as https://github.com/M0r13n/pyais/pull/148 is merged.
    """

    _fobj: socket

    def __init__(self, host: str, port: int) -> None:
        """Initialize the UDP receiver."""
        sock: socket = socket(AF_INET, SOCK_DGRAM)
        sock.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1)
        sock.bind((host, port))
        super().__init__(sock, preprocessor=None)

    def recv(self) -> bytes:
        """Receive data from socket."""
        return self._fobj.recvfrom(self.BUF_SIZE)[0]

    def close(self) -> None:
        """Close the UDP receiver."""
        self._fobj.close()


class AisTrackerCoordinator(
    DataUpdateCoordinator[dict[str, dict[str, float | int | str | None]]]
):
    """AIS tracker data update coordinator."""

    config_entry: AisTrackerConfigEntry

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the AIS tracker coordinator."""
        super().__init__(
            hass,
            LOGGER,
            name=DOMAIN,
            update_interval=None,
        )
        self._receiver: UDPReceiver | None = None
        self._background_task: asyncio.Task | None = None
        self.data = {}

    async def async_setup(self) -> None:
        """Set up the AIS tracker coordinator."""

        self._receiver = UDPReceiver("", self.config_entry.data[CONF_PORT])

        async def async_ais_listerner():
            def ais_listerner():
                for msg_raw in self._receiver:
                    msg = json_loads_object(msg_raw.decode().to_json())
                    LOGGER.debug("received msg: %s", msg)
                    if (
                        msg.get("msg_type") in [1, 2, 3, 5]
                        and str(msg.get("mmsi")) in self.config_entry.data[CONF_MMSIS]
                    ):
                        self.hass.create_task(self.async_process_ais_message(msg))

            await self.hass.async_add_executor_job(ais_listerner)

        self._background_task = self.config_entry.async_create_background_task(
            hass=self.hass,
            target=async_ais_listerner(),
            name=f"{DOMAIN}_ais_listerner",
            eager_start=False,
        )

        self.config_entry.async_on_unload(self.async_teardown)

    async def async_teardown(self) -> None:
        """Tear down the AIS tracker coordinator."""
        if self._receiver is not None:
            self._receiver.close()
        if self._background_task is not None:
            self._background_task.cancel()

    async def async_process_ais_message(self, message: dict) -> None:
        """Process incoming ais message."""
        LOGGER.debug("process ais message: %s", message)
        if message.get("msg_type") in [1, 2, 3]:  # position reports
            data = {**self.data}
            data[str(message["mmsi"])] = message
            self.async_set_updated_data(data)

        elif message.get("msg_type") == 5:  # Static and voyage related data
            dreg = dr.async_get(self.hass)
            dreg.async_get_or_create(
                config_entry_id=self.config_entry.entry_id,
                identifiers={(DOMAIN, str(message["mmsi"]))},
                name=f"{message.get('shipname')} ({message.get('callsign')})",
                serial_number=str(message["mmsi"]),
                manufacturer="AIS tracker",
            )
