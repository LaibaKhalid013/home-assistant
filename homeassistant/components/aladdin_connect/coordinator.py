"""Define an object to coordinate fetching Aladdin Connect data."""

from datetime import timedelta
import logging

from genie_partner_sdk.client import AladdinConnectClient
from genie_partner_sdk.model import GarageDoor

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


class AladdinConnectCoordinator(DataUpdateCoordinator[None]):
    """Aladdin Connect Data Update Coordinator."""

    doors: list[GarageDoor] | None = None

    def __init__(self, hass: HomeAssistant, acc: AladdinConnectClient) -> None:
        """Initialize."""
        super().__init__(
            hass,
            logger=_LOGGER,
            name=DOMAIN,
            update_interval=timedelta(seconds=15),
        )
        self.acc = acc

    async def async_setup(self) -> None:
        """Fetch initial data."""
        self.doors = await self.acc.get_doors()

    async def _async_update_data(self) -> None:
        """Fetch data from API endpoint."""
        assert self.doors is not None
        for door in self.doors:
            await self.acc.update_door(door.device_id, door.door_number)
