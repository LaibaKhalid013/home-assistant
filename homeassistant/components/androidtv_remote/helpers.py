"""Helper functions for Android TV Remote integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from androidtvremote2 import AndroidTVRemote

from homeassistant.helpers.storage import STORAGE_DIR

from .const import CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE

if TYPE_CHECKING:
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant


def create_api(hass: HomeAssistant, host: str, enable_ime: bool) -> AndroidTVRemote:
    """Create an AndroidTVRemote instance."""
    return AndroidTVRemote(
        client_name="Home Assistant",
        certfile=hass.config.path(STORAGE_DIR, "androidtv_remote_cert.pem"),
        keyfile=hass.config.path(STORAGE_DIR, "androidtv_remote_key.pem"),
        host=host,
        loop=hass.loop,
        enable_ime=enable_ime,
    )


def get_enable_ime(entry: ConfigEntry) -> bool:
    """Get value of enable_ime option or its default value."""
    return entry.options.get(CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE)
