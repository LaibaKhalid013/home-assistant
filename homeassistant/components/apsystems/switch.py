"""The power switch which can be toggled via the APsystems local API integration."""

from __future__ import annotations

from typing import Any

from homeassistant.components.switch import SwitchDeviceClass, SwitchEntity
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import DiscoveryInfoType

from . import ApSystemsConfigEntry, ApSystemsData
from .entity import ApSystemsEntity


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ApSystemsConfigEntry,
    add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up the switch platform."""

    add_entities([ApSystemsPowerSwitch(config_entry.runtime_data)])


class ApSystemsPowerSwitch(ApSystemsEntity, SwitchEntity):
    """The switch class for APSystems switches."""

    _attr_device_class = SwitchDeviceClass.SWITCH

    def __init__(
        self,
        data: ApSystemsData,
    ) -> None:
        """Initialize the switch."""
        super().__init__(data)
        self._api = data.coordinator.api
        self._attr_unique_id = f"{data.device_id}_power_switch"

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self._api.get_device_power_status() is not None

    @property
    def is_on(self) -> bool | None:
        """Return state of the switch."""
        return self._api.get_device_power_status()

    async def async_turn_on(self, **kwargs: Any) -> None:
        """Turn the switch on."""
        await self.hass.async_add_executor_job(self._api.set_device_power_status(0))
        await self._api.async_refresh()

    async def async_turn_off(self, **kwargs: Any) -> None:
        """Turn the switch off."""
        await self.hass.async_add_executor_job(self._api.set_device_power_status(1))
        await self._api.async_refresh()
