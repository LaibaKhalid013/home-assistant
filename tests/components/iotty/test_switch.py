"""Unit tests the Hass SWITCH component."""

from unittest.mock import AsyncMock

from aiohttp import ClientSession
from freezegun.api import FrozenDateTimeFactory
import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.iotty.api import IottyProxy
from homeassistant.components.iotty.const import DOMAIN
from homeassistant.components.iotty.coordinator import (
    UPDATE_INTERVAL,
    IottyData,
    IottyDataUpdateCoordinator,
)
from homeassistant.components.iotty.switch import IottyLightSwitch, async_setup_entry
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_entry_oauth2_flow
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .conftest import test_ls

from tests.common import MockConfigEntry, async_fire_time_changed


async def test_turn_on_ok(
    hass: HomeAssistant,
    mock_iotty: IottyProxy,
    mock_coordinator: IottyDataUpdateCoordinator,
    mock_iotty_command_fn,
) -> None:
    """Issue a turnon command."""
    mock_iotty.command = AsyncMock()
    sut = IottyLightSwitch(mock_coordinator, mock_iotty, test_ls[0])
    await sut.async_turn_on()
    await hass.async_block_till_done()

    mock_iotty.command.assert_called_once_with(
        test_ls[0].device_id, test_ls[0].cmd_turn_on()
    )


async def test_turn_off_ok(
    hass: HomeAssistant,
    mock_iotty: IottyProxy,
    mock_coordinator: IottyDataUpdateCoordinator,
    mock_iotty_command_fn,
) -> None:
    """Issue a turnoff command."""
    mock_iotty.command = AsyncMock()
    sut = IottyLightSwitch(mock_coordinator, mock_iotty, test_ls[0])
    await sut.async_turn_off()
    await hass.async_block_till_done()

    mock_iotty.command.assert_called_once_with(
        test_ls[0].device_id, test_ls[0].cmd_turn_off()
    )


async def test_setup_entry_wrongdomaindata_error(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_iotty: IottyProxy,
) -> None:
    """Setup the SWITCH entry with empty or wrong DOMAIN data."""

    with pytest.raises(KeyError):
        await async_setup_entry(hass, mock_config_entry, None)

    hass.data.setdefault(DOMAIN, {})
    with pytest.raises(KeyError):
        await async_setup_entry(hass, mock_config_entry, None)


async def test_setup_entry_ok_nodevices(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_async_add_entities: AddEntitiesCallback,
    local_oauth_impl: ClientSession,
    mock_coordinator: IottyDataUpdateCoordinator,
) -> None:
    """Correctly setup the SWITCH entry, with no iotty Devices to add to Hass."""

    mock_config_entry.add_to_hass(hass)

    config_entry_oauth2_flow.async_register_implementation(
        hass, DOMAIN, local_oauth_impl
    )

    hass.data.setdefault(DOMAIN, {})[mock_config_entry.entry_id] = mock_coordinator

    mock_coordinator.data = IottyData
    mock_coordinator.data.devices = {}

    await async_setup_entry(hass, mock_config_entry, mock_async_add_entities)

    assert len(mock_async_add_entities.mock_calls) == 1


async def test_setup_entry_ok_twodevices(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_async_add_entities: AddEntitiesCallback,
    local_oauth_impl: ClientSession,
    mock_coordinator: IottyDataUpdateCoordinator,
) -> None:
    """Correctly setup the SWITCH entry, with two iotty Devices to add to Hass."""

    mock_config_entry.add_to_hass(hass)

    config_entry_oauth2_flow.async_register_implementation(
        hass, DOMAIN, local_oauth_impl
    )

    hass.data.setdefault(DOMAIN, {})[mock_config_entry.entry_id] = mock_coordinator

    mock_coordinator.data = IottyData
    mock_coordinator.data.devices = test_ls
    mock_coordinator.iotty = IottyProxy

    await async_setup_entry(hass, mock_config_entry, mock_async_add_entities)

    assert len(mock_async_add_entities.mock_calls) == 1


async def test_devices_creaction_ok(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    local_oauth_impl: ClientSession,
    mock_get_devices_twolightswitches,
    mock_get_status_filled,
    snapshot: SnapshotAssertion,
) -> None:
    """Test iotty switch creation."""

    mock_config_entry.add_to_hass(hass)

    config_entry_oauth2_flow.async_register_implementation(
        hass, DOMAIN, local_oauth_impl
    )

    await hass.config_entries.async_setup(mock_config_entry.entry_id)

    assert hass.states.async_entity_ids() == snapshot


async def test_devices_deletion_ok(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    local_oauth_impl: ClientSession,
    mock_get_devices_twolightswitches_then_one,
    mock_get_status_filled,
    snapshot: SnapshotAssertion,
    freezer: FrozenDateTimeFactory,
) -> None:
    """Test iotty switch deletion."""

    mock_config_entry.add_to_hass(hass)

    config_entry_oauth2_flow.async_register_implementation(
        hass, DOMAIN, local_oauth_impl
    )

    await hass.config_entries.async_setup(mock_config_entry.entry_id)

    freezer.tick(UPDATE_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()

    assert hass.states.async_entity_ids() == snapshot
