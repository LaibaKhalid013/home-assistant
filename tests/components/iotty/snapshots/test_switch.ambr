# serializer version: 1
# name: test_devices_creaction_ok
  list([
    'switch.test_light_switch_0_test_serial_0_test_light_switch_0_test_serial_0',
    'switch.test_light_switch_1_test_serial_1_test_light_switch_1_test_serial_1',
  ])
# ---
# name: test_devices_deletion_ok
  list([
    'switch.test_light_switch_0_test_serial_0_test_light_switch_0_test_serial_0',
    'switch.test_light_switch_1_test_serial_1_test_light_switch_1_test_serial_1',
  ])
# ---
# name: test_devices_deletion_ok.1
  list([
    'switch.test_light_switch_0_test_serial_0_test_light_switch_0_test_serial_0',
  ])
# ---
# name: test_devices_insertion_ok
  list([
    'switch.test_light_switch_0_test_serial_0_test_light_switch_0_test_serial_0',
    'switch.test_light_switch_1_test_serial_1_test_light_switch_1_test_serial_1',
  ])
# ---
# name: test_devices_insertion_ok.1
  list([
    'switch.test_light_switch_0_test_serial_0_test_light_switch_0_test_serial_0',
    'switch.test_light_switch_1_test_serial_1_test_light_switch_1_test_serial_1',
    'switch.test_light_switch_2_test_serial_2_test_light_switch_2_test_serial_2',
  ])
# ---
# name: test_setup_entry_ok_nodevices
  list([
  ])
# ---
